### ISSUE TITLE
twisted.web._newclient.Response does not expect to have bodyDataFinished called during deliverBody_INITIAL

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8720]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)| @Lukasa reported|
|-|-|
|Trac ID|trac#8720|
|Type|defect|
|Created|2016-08-02 16:53:29Z|
|Branch|https://github.com/twisted/twisted/tree/8720-lukasa-responsestatechange|

Discovered while working on [#8489](https://github.com/twisted/twisted/issues/8489). Currently, twisted.web._newclient.Response expects that when it is created it will be handed a paused transport, and it can unpause it when a body-receiving protocol is passed to it.

However, when deliverBody is called it resumes the transport *first*, and then hooks up the body delivering protocol, and then finally transitions state. Three things can happen here:

1. calling resumeProducing does nothing. The protocol is hooked up, the state transitions, and then data starts arriving. This is the overwhelmingly likely case in the current arrangement, which expects a TCP transport.
2. calling resumeProducing delivers some small amount of data. This is ok: it will end up in the data buffer and be delivered to the protocol later in the method. No harm, no foul.
3. calling resumeProducing delivers a complete buffered response, including calling _bodyDataFinished. This does not work as expected, because the state has not transitioned yet. That means that Response ends up sitting around waiting for a call to _bodyDataFinished that will never come.

I believe this can be fixed by deferring the unpausing of the transport until *after* all the setup has been done and the state has transitioned, rather than by doing it at the top of the method.

<details><summary>Searchable metadata</summary>

```
trac-id__8720 8720
type__defect defect
reporter__Lukasa Lukasa
priority__normal normal
milestone__None None
branch__8720_lukasa_responsestatechange 8720-lukasa-responsestatechange
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__None None
time__1470156809647066 1470156809647066
changetime__1473297241222398 1473297241222398
version__None None
owner__glyph glyph

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_1"></a>|@Lukasa commented|
|-|-|

Annoyingly, moving the call to resumeProducing to the bottom of the method causes a failure in `test_abortAfterResponseHeaders`. Specifically, this failure is caused by the fact that StringTransport cannot be resumed after it has had connectionLost called on it. The body consumer in this case aborts the connection when `connectionMade` is called on it, which means that this order of events is *also* not acceptable.

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_2"></a>|@Lukasa commented|
|-|-|

We also cannot move the state transition to the top of the method and then immediately resume producing, because production resumption may also trigger data to be delivered, which would mean it got delivered first before the buffered data.

This seems to be a thorny issue: there doesn't seem to be anywhere safe for resumeProducing to go here.

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_3"></a>|@Lukasa commented|
|-|-|

Basically, the transport has to be resumed before makeConnection is called on the protocol, because makeConnection can abort the transport. The transport cannot be resumed until the state has transitioned to CONNECTED, because the resumption may call _bodyDataFinished. And the state cannot be transitioned to CONNECTED until after the buffered data is flushed, which cannot happen until makeConnection is called.

These three requirements are mutually exclusive: they cannot all happen at once.

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_4"></a>|@Lukasa commented|
|-|-|

I'm kind of at a loss here: the only way to get around this is to have a requirement that resumeProducing be unable to re-entrantly finish the body, which requires a pretty substantial refactor of the _newclient code to guarantee for the HTTP/2 case.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

Your assessment that moving the state transition to the top and only triggering the potentially re-entrant behavior afterwards is completely correct.  This is more or less how any async code _has_ to be written :).

If we take that as a given, then there are two problems here, which I think should be resolved separately:

1. `StringTransport` has a buggy implementation of `resumeProducing`.  It is buggy by definition, because the interfaces don't define any exceptions that can be raised, and if you want to be a formally correct implementation of an interface, you have to adhere to any `@raises` documentation it presents.  Perhaps more importantly, it's equally buggy by convention, because existing implementations don't raise this exception, and silently do nothing in this case.  The reason that I believe `StringTransport` has adopted a deliberately buggy approach here, I imagine, is to avoid silently letting _potentially_-erroneous state transitions pass silently; you want your tests to blow up if you did something that might silently have no effect in real production.  But in this case, that is deliberate.  So `StringTransport` should acquire a `lenient=False` constructor argument, or something along those lines, to disable this belt-and-suspenders behavior in cases like this one where it's unwanted.  If you change `StringTransport` in this way then you can give up on the requirement that "the transport has to be resumed" and the other remaining requirements are no longer mutually exclusive.
2. `Response` doesn't have good error handling in the case of a buggy `resumeProducing`.  That transport might be application-provided, and `Response` should probably handle [and log](https://twistedmatrix.com/documents/16.3.0/api/twisted.logger.Logger.html#failure) the error.  This is considerably lower priority though.

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_6"></a>|@Lukasa commented|
|-|-|

In changeset 361c55d17fddfc03317eec6e47f8e9170a336348
```
#!CommitTicketReference repository="" revision="361c55d17fddfc03317eec6e47f8e9170a336348"
Topfile for #8720.
```

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_7"></a>|@Lukasa commented|
|-|-|

See https://github.com/twisted/twisted/pull/465

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph set owner to @Lukasa|
|-|-|

Ugh. Can't believe it took me a month to get to this.  Conflicts have overtaken us in the meanwhile.  If you can resolve them, I will almost certainly be happy to merge, doubly so if you can address the few comments on the PR.  Please ping me when you've done so. Thanks, and sorry for the delay.

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_9"></a>|@Lukasa removed owner|
|-|-|

I believe I've addressed all the feedback left on Github.

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_10"></a>|@Lukasa commented|
|-|-|

In changeset 2fca626ade15622177fb1b8808bb583c24b87f65
```
#!CommitTicketReference repository="" revision="2fca626ade15622177fb1b8808bb583c24b87f65"
Fixup the topfile for #8720.
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph set owner to @glyph|
|-|-|

The changes look good to me, and I think we have an acceptable temporary workaround for the build failures.  Except, of course, Travis chose this moment to have an OS X build outage, so I'm waiting for that to come back around so we have the required build status for a merge.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph set status to `closed`|
|-|-|

In changeset c02c19441f6bd9d2431ac019ba36ac6abe33cfea
```
#!CommitTicketReference repository="" revision="c02c19441f6bd9d2431ac019ba36ac6abe33cfea"
Merge 8720-lukasa-responsestatechange: HTTP client state machine fix

Author: Lukasa

Reviewer: glyph

Fixes: #8720

twisted.web.client.HTTPConnectionPool and anything that uses it, like
twisted.web.client.Agent, have had their logic for resuming transports changed
so that transports are resumed after state machine transitions are complete,
rather than before. This change allows the HTTP client infrastructure to work
with alternative HTTP implementations such as HTTP/2 which may be able to
deliver a complete response synchronously when producing is resumed.
```
